docs/when_internals.md

```diff
@@ -0,0 +1,342 @@
+    are any latches
+  * `when`, `elsewhen`, `otherwise`: public APIs to interact with the when context
+    global state to construct corresponding objects tracking the logic
+
+
+* `wire_container.py`
+  * `Wireable`
+    * Overview
+      * The main `when` logic for the `Wireable` class is to trigger the `when`
+        code path when `wire` is invoked inside an active `when` context.  It is
+        also responsible for handling unwiring, as well as special logic for
+        wiring the output of a when primitive.
+    * Attributes:
+      * `self._enclosing_when_context`: tracks the active when context when a
+        `Wireable` value is created.  This ensures that wiring done inside the
+        enclosing context is treated as unconditional.  For example, if we have
+        the following code:
+        ```python
+        with m.when(io.x):
+            io.O @= ~io.I
+        ```
+        the `Invert` instance input port created for the `~` operator should be
+        unconditionally wired by `io.I` since it is created inside the context.
+      * `self._wire_when_contexts`: tracks the when contexts in which this value
+        was wired, which provides faster lookup than having to traverse the
+        possible contexts to search for the value.
+      * `self._is_when_port`: flag used to track whether a value was create by a
+        when builder instance, which enables a special code path for when a when
+        output value is directly wired.  This can occur when a user calls
+        `.value()` on an conditionally driven value and is returned a handle to
+        the when primitive output.
+    * Properties
+      * `enclosing_when_context`: Set by recursive values (i.e. Tuples and
+        Arrays) when constructing child references (i.e. children inherit the
+        enclosing context of the parent).
+    * Methods
+      * `remove_from_wire_when_contexts`: used in the `unwire` logic for
+        conditionally driven values
+      * `unwire`: has an optional argument `keep_wire_when_contexts` that
+        preserves the conditional driver info.  This is used when a value
+        is driven by multiple when contexts, where the default value
+        may come from a previous when context.  In this case, the value will
+        be rewired to new output of the when primitive, but it will maintain
+        references to the previous contexts in which it was driven.
+      * `_when_output_wire`: Logic responsible for ensuring invariants when
+        a user directly wires the output of a when primitive (usually
+        referenced by valling `.value()` on a conditionally driven value.).
+        This code effectively copies the conditional drivers of the driven
+        value and adds them to the same when contexts as drivers for the 
+        a value.  This ensures that a when output only drives one value,
+        so when a user attempts to wire a when output, we create a new,
+        identical when output to drive the target value.
```

Comment:
We could add a temporary value here shared by the drivees which would reduce code in the output (share the assignments in the when)